src/hydration/hydration.ts

```diff
@@ -0,0 +1,127 @@
+import { DEFAULT_STALE_TIME, DEFAULT_CACHE_TIME } from '../core/config'
+
+import type { Query, QueryCache, QueryKey, QueryConfig } from 'react-query'
+
+export interface DehydratedQueryConfig<TResult> {
+  staleTime?: number
+  cacheTime?: number
+  initialData?: TResult
+}
+
+export interface DehydratedQuery<TResult> {
+  config: DehydratedQueryConfig<TResult>
+  updatedAt: number
+}
+
+export interface DehydratedQueries<TResult> {
+  [hash: string]: DehydratedQuery<TResult>
+}
+
+export type QueryKeyParserFunction = (queryHash: string) => QueryKey
+export type ShouldHydrateFunction = <TResult>({
```

Comment:
Should we provide this functionality? Maybe better to wait until someone has a request or use-case for it, otherwise we might need to support functionality which nobody uses

Reply:
Great question! I think we do, this came up in a discussion with Tanner and the use case is avoiding hydrating queries that have become "extra stale" kind of, that is, they are too old to even show the data while refetching in the background. We only know that at hydration time. Makes sense?

This is there for persisting to and from localstorage, not SSR, since using it with SSR would mean hydration mismatches.

Reply:
Maybe this use case is not hydration specific? For example: if the user is on a page and then after a while navigates to another page, then the data might also be too old to show there. The existence of a query in the cache should not really matter.

Reply:
Isn't `cacheTime` there for that on the client? Since we don't want to make that based on `updatedAt` like `staleTime`, we need something else for hydration.

Reply:
The GC timer would not have started because the previous page was still showing it, but the data might be too old to show on the next page. You do bring up a good point though, should queries be hydrated when the cache time exceeded?

Reply:
Ah yes, I think there are probably other good arguments to be made that `cacheTime` is slightly different as well.

We do need to hydrate queries even if `cacheTime` is exceeded, or we will get React hydration mismatch warnings with SSR. :)

Reply:
I removed shouldHydrate for now. We can always add it back, or add the more generic mechanism later.